src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+                where T1 : IDisposable where T2 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second }, false);
+                return (first, second);
+            }
+
+            public (T1 first, T2 second, T3 third) Exclude<T1, T2, T3>(T1 first, T2 second, T3 third)
+                where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second, third }, false);
+                return (first, second, third);
+            }
+
+            public IDisposable[] Exclude(IDisposable[] exclude, bool excludeGlobally)
+            {
+                lock (this) {
+                    foreach (var disposable in exclude) {
+                        if (Disposables.Contains(disposable)) {
+                            Disposables.Remove(disposable);
+                            if (!excludeGlobally) {
+                                Singleton.MoveToOuterScope(this, disposable);
+                            }
+                        }
+                    }
+                }
+
+                return exclude;
+            }
+
+            /// <summary>
+            /// Excludes a set of tensors/disposables from the all dispose scopes, see overloaded methods. See Exclude
+            /// if you wish to move it to the outer dispose scope.
+            /// </summary>
+            public T ExcludeGlobally<T>(T exclude) where T : IDisposable
+            {
+                Exclude(new IDisposable[] { exclude }, true);
+                return exclude;
+            }
+
+            public (T1 first, T2 second) ExcludeGlobally<T1, T2>(T1 first, T2 second)
+                where T1 : IDisposable where T2 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second }, true);
+                return (first, second);
+            }
+
+            public (T1 first, T2 second, T3 third) ExcludeGlobally<T1, T2, T3>(T1 first, T2 second, T3 third)
+                where T1 : IDisposable where T2 : IDisposable where T3 : IDisposable
+            {
+                Exclude(new IDisposable[] { first, second, third }, true);
+                return (first, second, third);
+            }
+
+            /// <summary>
+            /// Disposes everything currenly in the dispose scope.
+            /// </summary>
+            public void DisposeEverything() => DisposeEverythingBut();
+
+            /// <summary>
+            /// As an intermediate step, you can dispose all the tensors/disposables currently scheduled for dispose, to
+            /// clear up some memory without creating a new scope. Note that this doesn't permanently exclude the
+            /// tensors from disposing, use Exclude for that. Also, excluded tensors don't need to be included
+            /// here.
+            /// </summary>
+            public IDisposable[] DisposeEverythingBut(params IDisposable[] keep)
```

Comment:
I've found params arrays super-useful in TorchSharp, but I've also found it super-useful to have an IEnumerable<T> version and put the implementation inside there. Sometimes, it's just so useful to build up a data structure dynamically and pass it in, and to being able to avoid calling 'ToList()' and make a copy would be ideal.

Reply:
It's because of the params parameter, you can't use IEnumberable<IDisposable> as a params parameter. I'm not sure when  someone would use that fallback, though, so an IEnumerable version might suffice 99.9% of the time.

Reply:
I changed both 

public void DisposeEverythingBut(IEnumerable<IDisposable> keep)

and

  public void Exclude(IEnumerable<IDisposable> exclude, bool excludeGlobally)